src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1917,6 +1917,15 @@
    This sums to 65,527B just shy of the 65,536B method size limit."
   5000)
 
+(def retain-compile-ctx-default
+  "During construction of the Session there is data maintained such that if the underlying expressions fail to compile
+   then this data can be used to explain the failure and the constraints of the rule whos expression is being evaluated.
+   The default behavior will be to discard this data, as there will be no use unless the session will be serialized and
+   deserialized into a dissimilar environment, ie function or symbols might be unresolvable. In those sorts of scenarios
+   it would be possible to construct the original Session with the `retain-compile-ctx` flag set to true, then the compile
+   context should aid in debugging the compilation failure on deserialization."
+  false)
```

Comment:
My initial instinct here would actually be that the default should be true, and only users who need to be concerned with memory optimizations on huge rulesets would be concerned with this.  That's a debatable point though.

Also, would it be possible to include a smaller set of information rather than the entire contents?  Would that achieve your memory use goals without introducing the complexity of a new user-facing setting?  I see [here](https://github.com/cerner/clara-rules/blob/0.19.0/src/main/clojure/clara/rules/compiler.clj#L1262) that it looks like we're binding the production to the compile-ctx dynamic var, perhaps we could retain something smaller like the name instead of the entire structure?

Reply:
I felt that it should default to false as the enhancement to batched compilation #396 could actually be seen as a degradation of memory for the typical use-cases (assuming non-durability is typical). Before this information was discarded right after evaluation, this way it would be putting it back the way it was before without requiring memory conscious consumers to have to evaluate the addition of the flag. I'm ok with it either way.

I'm not to worried about bindings that would fall out of scope after the Session has been created, rather i am trying to avoid the session itself holding this compile-ctx after it has been compiled.

Reply:
Discussed offline some with @EthanEChristian - I think the primary problem here is that the idea of a "compile-ctx" is something of a Clara/Clojure implementation detail that we shouldn't force users to dedicate thought to.  If we rename this ":deserialization-dbg-mode?" though, or something similar that more directly corresponds to the decision users need to make and document it accordingly, that seems like a reasonable tradeoff to me.  We'd preferably also add a note in the exceptions thrown if the extra data isn't present reminding the user that this mode exists i.e. "An exception was thrown while deserialization the rulebase.  Note that extra information is available when the :deserialization-dbg-mode is enabled".

Reply:
I'm not a fan of how specific `deserialization-dbg-mode?` is when the data stored/not-stored is more generic "compilation context".

I actually think the name chosen seems appropriate, ie `retain-compile-ctx`

The only odd part to me is that it isn't stated in reverse perhaps, ie. `omit-compile-ctx` (or as clj/core likes to say "elide"). This would default to true. I think "omit" seems to be more common direction when referring to compiler opts - sort of like some opts in cljs.

Reply:
My primary concern here isn't exactly the name in and of itself, but that Clara be accessible and as easy to comprehend as possible for users who aren't familiar with Clojure or its ecosystem.  So for example, ideally a user need not even be aware that Clara makes calls to clojure.core/eval behind the scenes, or even of clojure.core/eval's existence.  My concern is that the current way this option appears makes users think they may need to be concerned with it when that is not in fact the case.  Perhaps we could reword the session options docs to something like ""retain compile context: When true Clara in Clojure retains additional information to improve error messages during session deserialization at the cost of additional memory use.  ... [rest of docs]. " Basically I'd prefer to make it easy for users only skimming the docstrings to quickly determine what settings concern them.  I think we can retain the name if preferred while still accomplishing that goal.

Reply:
Updated the message in the latest commit, also changed the flag to `omit` per a comment by @mrrodriguez